File: Backend/src/Database/models/hotelModel.ts
import mongoose, { Schema } from "mongoose";
import { IHotel } from "../DTOs/HotelsDTO";

const hotelSchema: IHotel = new Schema<IHotel>({
    hotel_name: {
        type: String,
        required: true
    },
    description: {
        type: String,
        required: true
    },
    category: {
        type: String,
        required: true,
        enum: ["Hotel", "Resort", "Apartment", "Villa", "Hostel"]
    },
    starRating: {
        type: Number,
        required: true,
        min: 1,
        max: 5
    },
    location: {
        address: {
            type: String,
            required: true
        },
        city: {
            type: String,
            required: true
        },
        state: {
            type: String,
            required: true
        },
        country: {
            type: String,
            required: true
        },
        zipCode: {
            type: String,
            required: true
        },
        geo: {
            type: {
                type: String,
                enum: ['Point'],
                default: 'Point'
            },
            coordinates: {
                type: [Number],
                index: '2dsphere'
            }
        }
    },
    contact: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'Contact'
    },
    images: [
        {
            cloudinaryURL: {
                type: String
            },
            cloudinaryPublicId: {
                type: String
            }
        }
    ],
    amenities: [String],
    checkInTime: {
        type: String,
        default: '14:00'
    },
    checkOutTime: {
        type: String,
        default: "11:00"
    },
    rooms: [
        {
            type: mongoose.Schema.Types.ObjectId,
            ref: 'Room'
        }
    ],
    reviews: [
        {
            userId: {
                type: mongoose.Schema.Types.ObjectId,
                ref: 'User'
            },
            userName: {
                type: String
            },
            rating: {
                type: Number,
                min: 1,
                max: 5
            },
            comment: {
                type: String
            }
        }
    ]
}, {timestamps: true});

const Hotel = mongoose.model('Hotel', hotelSchema);

export default Hotel;

File: APIs/app/database/models/hotel_model.py
from beanie import Document, Link, PydanticObjectId
from pydantic import Field, BaseModel
from typing import List, Optional
from enum import Enum
from .contact_model import Contact
from .room_model import Room
from .review_model import Review

class HotelCategory(str, Enum):
    Hotel = "Hotel"
    Resort = "Resort"
    Apartment = "Apartment"
    Villa = "Villa"
    Hostel = "Hostel"

class GeoLocation(BaseModel):
    type: str = "Point"
    coordinates: List[float] # [Longitude, Latitude]

class Location(BaseModel):
    address: str
    city: str
    state: str
    country: str
    zipCode: str
    geo: Optional[GeoLocation] = None

class HotelImage(BaseModel):
    cloudinaryURL: Optional[str] = None
    cloudinaryPublicId: Optional[str] = None

class Hotel(Document):
    hotel_name: str
    description: str
    category: HotelCategory
    starRating: int = Field(..., ge=1, le=5)
    location: Location
    contact: Optional[Link[Contact]] = None
    images: List[HotelImage] = []
    amenities: List[str] = []
    checkInTime: str = "14:00"
    checkOutTime: str = "11:00"
    rooms: List[Link[Room]] = []
    reviews: List[Link[Review]] = []

    class Settings:
        name = "hotels"
